src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1080,7 +1080,7 @@ public override IEnumerable<string> MemberNames
         {
             get
             {
-                return IsTupleType ? GetMembers().Select(m => m.Name).Distinct() : this.declaration.MemberNames;
+                return (IsTupleType || IsRecord) ? GetMembers().Select(m => m.Name).Distinct() : this.declaration.MemberNames;
```

Comment:
:memo: I'm debating whether we  should add caching of `MemberNames` in `SourceMemberContainerSymbol` at this point... Thoughts?

Reply:
> I'm debating whether we should add caching of MemberNames in SourceMemberContainerSymbol at this point... Thoughts?

I would prefer that... but it's not necessarily a huge deal if you don't (unlesss people have tons of records).

The purpose of these collectoins is to allow the IDE to search for names very efficiently (i.e. in a non-allocating fashion).  So it is important that *the vast majority of source types* not allocate here.  And, as long as that holds we'll be ok.

But it would be a bit nicer to cache IMO in case you had someone with, say, 1k generated records, and we kept generaitng this over and over again.  

Note: You don't need to do Distinct.  MemberNames is *explicitly* doc'ed as returning dupes.  exactly because we don't want to do the CPU/memory work to find and remove duplicates for htese sensitive scenarios.

Reply:
> MemberNames is explicitly doc'ed as returning dupes

Where do you see that?
The implementations I looked at (such as named type symbols) seem to explicitly de-dupe.

Reply:
```c#
        public ICollection<string> MemberNames
        {
            get
            {
                if (_lazyMemberNames == null)
                {
                    var names = UnionCollection<string>.Create(this.Declarations, d => d.MemberNames);
                    Interlocked.CompareExchange(ref _lazyMemberNames, names, null);
                }

                return _lazyMemberNames;
            }
        }
```

The merged decl for named types (i.e. for partials) just literally 'combines' (i.e. concatenates) all the names of all the individual parts. SourceMemberContainer just does this:

```
        public override IEnumerable<string> MemberNames
        {
            get
            {
                return IsTupleType ? GetMembers().Select(m => m.Name).Distinct() : this.declaration.MemberNames;
            }
        }
```

So this will return dupes.  We do not do any sort of 'distinct' as we want this not to cost anything in terms of memory/CPU.  This was very intentional from teh impl.  Source: i wrote it :)

Reply:
@CyrusNajmabadi Thanks! Removed the `.Distinct` :-)